main.c

```diff
@@ -74,12 +79,129 @@ void usage(int error)
 	fprintf(dest, "        -v               : Verbose\n");
 	fprintf(dest, "        -s               : Signed using bootsig.bin\n");
 	fprintf(dest, "        -0/1/2/3/4/5/6   : Only look for CMs attached to USB port number 0-6\n");
-	fprintf(dest, "        -p [pathname]    : Only look for CM with USB pathname\n");
+	fprintf(dest, "        -p [pathname]    : Only look for CM with USB pathname\n");\
+	fprintf(dest, "        -i [serialno]    : Only look for a Raspberry Pi Device with a given serialno\n");
 	fprintf(dest, "        -h               : This help\n");
 
 	exit(error ? -1 : 0);
 }
 
+libusb_device_handle * LIBUSB_CALL open_device_with_serialno(
+	libusb_context *ctx, char* serialno
+) {
+	struct libusb_device **devices;
+	struct libusb_device *cursor;
+	struct libusb_device_handle *handle = NULL;
+	int r = 0;
+
+	if (libusb_get_device_list(ctx, &devices) < 0)
+		return NULL;
+
+	uint32_t device_index = 0;
+	unsigned char *serial_buffer = calloc(33, sizeof(uint8_t));
+	if (serial_buffer == NULL) {
+		goto out_serialno;
+	}
+
+	while ((cursor = devices[device_index++]) != NULL) {
+		struct libusb_device_descriptor desc;
+		r = libusb_get_device_descriptor(cursor, &desc);
+		if (r < 0) {
+			goto out_serialno;
+		}
+
+		r = libusb_open(cursor, &handle);
+		if (r < 0) {
+			goto out_serialno;
+		}
+		
+		if (handle != NULL) {
+			if (libusb_get_string_descriptor_ascii(handle, desc.iSerialNumber, serial_buffer, 31) >= 0) {
+				if (strncmp(serialno, (char *)serial_buffer, 32)) {
+					handle = NULL;
+					continue;
+				}
+				
+				// Match the magic numbers for Raspberry Pi generations
+				if (desc.idVendor == 0x05ac) {
```

Comment:
Could potentially factor this out - but wanted to make a minimal change to existing code.